pilot/pkg/bootstrap/namespacecontroller.go

```diff
@@ -51,83 +51,77 @@ type NamespaceController struct {
 func NewNamespaceController(data func() map[string]string, kubeClient kubernetes.Interface) *NamespaceController {
 	c := &NamespaceController{
 		getData: data,
 		client:  kubeClient.CoreV1(),
-		queue:   queue.NewQueue(time.Second),
 	}
 
-	configmapInformer := informer.NewFilteredConfigMapInformer(kubeClient, metav1.NamespaceAll, 0,
-		cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc},
-		func(options *metav1.ListOptions) {
-			options.LabelSelector = fields.SelectorFromSet(configMapLabel).String()
-		})
-	configmapInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
-		UpdateFunc: func(oldObj, newObj interface{}) {
-			c.queue.Push(func() error {
-				return c.configMapChange(newObj)
-			})
-		},
-		DeleteFunc: func(obj interface{}) {
-			cm, ok := obj.(*v1.ConfigMap)
-			if !ok {
-				tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
-				if !ok {
-					log.Errorf("error decoding object, invalid type")
-					return
-				}
-				cm, ok = tombstone.Obj.(*v1.ConfigMap)
-				if !ok {
-					log.Errorf("error decoding object tombstone, invalid type")
-					return
-				}
-			}
-			c.queue.Push(func() error {
-				ns, err := kubeClient.CoreV1().Namespaces().Get(cm.Namespace, metav1.GetOptions{})
-				if err != nil {
-					return err
-				}
-				// If the namespace is terminating, we may get into a loop of trying to re-add the configmap back
-				// We should make sure the namespace still exists
-				if ns.Status.Phase != v1.NamespaceTerminating {
-					return c.insertDataForNamespace(cm.Namespace)
-				}
-				return nil
-			})
-		},
-	})
-	c.configMapController = configmapInformer
-
-	namespaceInformer := informer.NewNamespaceInformer(kubeClient, 0, cache.Indexers{})
-	namespaceInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
-		AddFunc: func(obj interface{}) {
-			c.queue.Push(func() error {
-				return c.namespaceChange(obj)
-			})
-		},
+	namespaceLW := listwatch.MultiNamespaceListerWatcher([]string{metav1.NamespaceAll}, func(namespace string) cache.ListerWatcher {
+		return &cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				return c.client.Namespaces().List(options)
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				return c.client.Namespaces().Watch(options)
+			}}
 	})
-	c.namespaceController = namespaceInformer
+
+	c.namespaceStore, c.namespaceController =
+		cache.NewInformer(namespaceLW, &v1.Namespace{}, namespaceResyncPeriod, cache.ResourceEventHandlerFuncs{
+			UpdateFunc: c.namespaceUpdated,
+			AddFunc:    c.namespaceAdded,
+		})
 
 	return c
 }
 
+// When a namespace is created, Citadel adds its public CA certificate
+// to a well known configmap in the namespace.
+func (nc *NamespaceController) namespaceAdded(obj interface{}) {
+	ns, ok := obj.(*v1.Namespace)
+
+	if ok {
+		err := nc.insertDataForNamespace(ns.Name)
+		if err != nil {
+			log.Errorf("error when inserting CA cert to configmap: %v", err)
```

Comment:
error is not handled so any transient error will lead to permanently missing updates to namespace. That is why we had the queue

Reply:
Any transient error will be handled later by the update event handler.